src/core/backbone_router/bbr_manager.cpp

```diff
@@ -119,39 +142,70 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
     VerifyOrExit(addressesLength / sizeof(Ip6::Address) <= kIPv6AddressesNumMax,
                  status = ThreadStatusTlv::kMlrGeneralFailure);
 
-    IgnoreError(Get<BackboneRouter::Leader>().GetConfig(config));
-    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(config.mMlrTimeout);
-
-    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+    if (!processTimeoutTlv)
     {
-        bool failed = true;
+        IgnoreError(Get<BackboneRouter::Leader>().GetConfig(config));
 
-        IgnoreReturnValue(aMessage.Read(addressesOffset + offset, sizeof(Ip6::Address), &address));
+        timeout = config.mMlrTimeout;
+    }
+    else
+    {
+        VerifyOrExit(timeout < std::numeric_limits<uint32_t>::max(), status = ThreadStatusTlv::kMlrNoPersistent);
 
-        switch (mMulticastListenersTable.Add(address, expireTime))
+        if (timeout != 0)
         {
-        case OT_ERROR_NONE:
-            failed = false;
-            break;
-        case OT_ERROR_INVALID_ARGS:
-            if (status == ThreadStatusTlv::kMlrSuccess)
-            {
-                status = ThreadStatusTlv::kMlrInvalid;
-            }
-            break;
-        case OT_ERROR_NO_BUFS:
-            if (status == ThreadStatusTlv::kMlrSuccess)
+            uint32_t origTimeout = timeout;
+
+            timeout = std::min(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMax));
+            timeout = std::max(timeout, static_cast<uint32_t>(Mle::kMlrTimeoutMin));
+
+            if (timeout != origTimeout)
             {
-                status = ThreadStatusTlv::kMlrNoResources;
+                otLogNoteBbr("MLR.req: MLR timeout is normalized from %u to %u", origTimeout, timeout);
             }
-            break;
-        default:
-            OT_ASSERT(false);
         }
+    }
+
+    expireTime = TimerMilli::GetNow() + TimeMilli::SecToMsec(timeout);
 
-        if (failed)
+    for (uint16_t offset = 0; offset < addressesLength; offset += sizeof(Ip6::Address))
+    {
+        IgnoreReturnValue(aMessage.Read(addressesOffset + offset, sizeof(Ip6::Address), &address) ==
```

Comment:
Do you mean `VerifyOrExit`?

Reply:
We already made sure there is enough data to read. 
But should revise to 
```c++
IgnoreReturnValue(aMessage.Read(addressesOffset + offset, sizeof(Ip6::Address), &address));
```